src/haliax/nn/linear.py

```diff
@@ -124,3 +129,134 @@ def _out_first(self):
+
+    @staticmethod
+    def init(
+        Experts: AxisSpec,
+        In: AxisSpec,
+        Out: AxisSpec,
+        *,
+        key: PRNGKey,
+        use_bias: bool = True,
+        out_first: bool = False,
+        init_scale: float = 1.0,
+    ) -> "Linear":
+        """
+        Args:
+            In: AxisSpec: The input axis spec
+            Out: AxisSpec: The output axis spec
+            key: PRNGKeyArray: The PRNG key to use for initialization
+            use_bias: bool: Whether to use a bias term
+            out_first: bool: Whether to put output axes first in the weight matrix. out_first is how PyTorch does it.
+            dot_general: Callable: The dot_general function to use. Defaults to jax.lax.dot_general. For fp8 or int8
+            init_scale: float: The scale to use for initialization. We scale init by 1/sqrt(Input.size)*init_scale
+        """
+
+        # TODO: naive implementation only supports single axis mapping for now
+        assert isinstance(Experts, Axis)
+        assert isinstance(In, Axis)
+        assert isinstance(Out, Axis)
+
+        joint_spec = hax.concat_axis_specs(Out, In) if out_first else hax.concat_axis_specs(In, Out)
+        joint_spec = hax.concat_axis_specs(Experts, joint_spec)
+        input_size = hax.axis_size(In)
+        weight = hax.random.truncated_normal(key, joint_spec, -3, 3) * (init_scale / math.sqrt(input_size))
+        bias = hax.zeros(Out) if use_bias else None
+
+        return MoELinear(weight, bias, Experts, In, Out)
+
+    @named_call
+    def __call__(self, inputs, group_sizes, *, key: Optional[PRNGKey] = None):
+        """
+        Args:
+            inputs (NamedArray): Input array    (Batch, Hidden)
+            group_sizes (NamedArray): MoE expert sizes (Experts)
+            key: Not used, but there for compat with other modules
+        """
+        del key
+
+        out_axes = inputs.axes[:-1] + (self.Out,)
+
+        q = _gmm(
+            inputs,
+            self.weight,
+            group_sizes,
+        )  # gmm((B, D), (E, D, d)) -> (B, d)
+        q = hax.NamedArray(q, axes=out_axes)
+        q = hax.auto_sharded(q)
+
+        if self.bias is not None:
+            q = q + self.bias
+            q = hax.auto_sharded(q)
+
+        return q
+
+    @property
+    def out_first(self):
+        """
+        Returns: bool: Whether the output axes are first in the weight matrix
+        """
+        # We do it this way because of scan layers
+        if isinstance(self.Out, hax.Axis):
+            return self.weight.axes[-1] != self.Out
+        else:
+            return self.weight.axes[-len(self.Out) :] != self.Out
+
+
+def _gmm(lhs, rhs, group_sizes, sharded=False):
+    if sharded:
+        gmm_fn = gmm_sharded
+    else:
+        gmm_fn = shard_map(
+            gmm_sharded,
+            mesh=hax.partitioning._get_mesh(),
+            in_specs=(
+                hax.partitioning.pspec_for_axis(lhs.axes),
```

Comment:
I think we want to force the all-gather. I think the right thing to do is to make a new axis mapping that deshards self.In, and deshard that axis mapping.